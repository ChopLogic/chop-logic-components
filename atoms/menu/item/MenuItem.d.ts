import { ChopLogicOrientationMode } from '../../../enums';
import { ChopLogicMenuItem } from '../../../models';
import { default as React, PropsWithChildren } from '../../../../node_modules/react';
type Props = PropsWithChildren & {
    item: ChopLogicMenuItem;
    mode: ChopLogicOrientationMode;
    openedOn?: 'hover' | 'click';
};
export declare const MenuItem: React.FC<Props>;
export {};
