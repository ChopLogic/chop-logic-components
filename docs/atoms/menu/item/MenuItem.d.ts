import { default as React, PropsWithChildren } from '../../../../node_modules/react';
import { ChopLogicOrientationMode } from '../../../enums';
import { ChopLogicMenuItem } from '../../../models';
type Props = PropsWithChildren & {
    item: ChopLogicMenuItem;
    mode: ChopLogicOrientationMode;
    openedOn?: 'hover' | 'click';
};
export declare const MenuItem: React.FC<Props>;
export {};
