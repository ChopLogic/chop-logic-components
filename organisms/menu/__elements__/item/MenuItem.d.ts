import { ChopLogicOrientationMode } from '../../../../enums';
import { ChopLogicMenuItem } from '../../../../models';
import { FC, PropsWithChildren } from '../../../../../node_modules/react';
type Props = PropsWithChildren & {
    item: ChopLogicMenuItem;
    mode: ChopLogicOrientationMode;
    openedOn?: 'hover' | 'click';
};
export declare const MenuItem: FC<Props>;
export {};
