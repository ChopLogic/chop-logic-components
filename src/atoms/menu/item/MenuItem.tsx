import { MenuLeaf } from '@atoms/menu/leaf/MenuLeaf';
import { ChopLogicOrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import { getClassName } from '@utils';
import React, { PropsWithChildren, useState } from 'react';

import { SubMenu } from '../sub-menu/SubMenu';
import styles from './MenuItem.module.scss';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  mode: ChopLogicOrientationMode;
  openedOn?: 'hover' | 'click';
};

export const MenuItem: React.FC<Props> = ({ item, mode, openedOn }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([
    styles.subMenuBar,
    { [styles.subMenuBar__horizontal]: mode === ChopLogicOrientationMode.Horizontal },
  ]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  const openSubMenu = () => {
    setIsSubMenuOpened(true);
  };

  return (
    <SubMenu
      item={item}
      isSubMenuOpened={isSubMenuOpened}
      toggleSubMenu={toggleSubMenu}
      closeSubMenu={closeSubMenu}
      openSubMenu={openSubMenu}
      mode={mode}
      openedOn={openedOn}
    >
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuItem item={child} key={child.id} mode={mode} openedOn={openedOn} />)}
      </ul>
    </SubMenu>
  );
};
