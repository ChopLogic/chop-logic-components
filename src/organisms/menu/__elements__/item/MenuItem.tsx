import { ChopLogicOrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import { getClassName } from '@utils';
import { FC, PropsWithChildren, useState } from 'react';

import { MenuLeaf } from '../leaf/MenuLeaf';
import { SubMenu } from '../sub-menu/SubMenu';
import styles from './MenuItem.module.scss';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  mode: ChopLogicOrientationMode;
  openedOn?: 'hover' | 'click';
};

export const MenuItem: FC<Props> = ({ item, mode, openedOn }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([
    styles.subMenuBar,
    { [styles.subMenuBar__horizontal]: mode === ChopLogicOrientationMode.Horizontal },
  ]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  const openSubMenu = () => {
    setIsSubMenuOpened(true);
  };

  return (
    <SubMenu
      item={item}
      isSubMenuOpened={isSubMenuOpened}
      toggleSubMenu={toggleSubMenu}
      closeSubMenu={closeSubMenu}
      openSubMenu={openSubMenu}
      mode={mode}
      openedOn={openedOn}
    >
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuItem item={child} key={child.id} mode={mode} openedOn={openedOn} />)}
      </ul>
    </SubMenu>
  );
};
