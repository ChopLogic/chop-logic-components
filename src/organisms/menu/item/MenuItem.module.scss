@use '/src/styles/animations/index';
@use '/src/styles/variables/index';
@use '/src/styles/mixins/index';

.subMenuBar {
  @include mixins.menuBar();
  @include animations.fade-in();

  flex-direction: column;
  padding-left: var(--cl-l-gap);
  margin-left: var(--cl-l-gap);
  margin-top: var(--cl-s-gap);
  box-sizing: border-box;
  border-left: var(--cl-primary-border);

  &__horizontal {
    position: absolute;
    top: 100%;
    left: 40%;
    width: max-content;
    border-radius: var(--cl-border-radius);
    border: var(--cl-primary-border);
    box-shadow: var(--cl-box-shadow);
    background-color: var(--cl-primary-background);
    padding: var(--cl-m-gap);
    margin: 0;
    z-index: variables.$popup;
  }
}
