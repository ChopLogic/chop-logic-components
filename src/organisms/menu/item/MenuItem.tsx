import { OrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import { getClassName } from '@utils';
import { FC, PropsWithChildren, useState } from 'react';

import { MenuLeaf } from '../leaf/MenuLeaf';
import { SubMenu } from '../sub-menu/SubMenu';
import styles from './MenuItem.module.scss';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  mode: OrientationMode;
  openedOn?: 'hover' | 'click';
};

export const MenuItem: FC<Props> = ({ item, mode, openedOn }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([styles.subMenuBar, { [styles.subMenuBar__horizontal]: mode === OrientationMode.Horizontal }]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  const openSubMenu = () => {
    setIsSubMenuOpened(true);
  };

  return (
    <SubMenu
      item={item}
      isSubMenuOpened={isSubMenuOpened}
      toggleSubMenu={toggleSubMenu}
      closeSubMenu={closeSubMenu}
      openSubMenu={openSubMenu}
      mode={mode}
      openedOn={openedOn}
    >
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuItem item={child} key={child.id} mode={mode} openedOn={openedOn} />)}
      </ul>
    </SubMenu>
  );
};
