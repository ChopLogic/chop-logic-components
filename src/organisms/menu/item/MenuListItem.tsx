import { OrientationMode } from '@enums';
import { MenuItem } from '@models';
import { getClassName } from '@utils';
import { FC, PropsWithChildren, useState } from 'react';

import { MenuLeaf } from '../leaf/MenuLeaf';
import { SubMenu } from '../sub-menu/SubMenu';
import styles from './MenuListItem.module.scss';

type Props = PropsWithChildren & {
  item: MenuItem;
  mode: OrientationMode;
  openedOn?: 'hover' | 'click';
};

export const MenuListItem: FC<Props> = ({ item, mode, openedOn }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([styles.subMenuBar, { [styles.subMenuBar__horizontal]: mode === OrientationMode.Horizontal }]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  const openSubMenu = () => {
    setIsSubMenuOpened(true);
  };

  return (
    <SubMenu
      item={item}
      isSubMenuOpened={isSubMenuOpened}
      toggleSubMenu={toggleSubMenu}
      closeSubMenu={closeSubMenu}
      openSubMenu={openSubMenu}
      mode={mode}
      openedOn={openedOn}
    >
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuListItem item={child} key={child.id} mode={mode} openedOn={openedOn} />)}
      </ul>
    </SubMenu>
  );
};
