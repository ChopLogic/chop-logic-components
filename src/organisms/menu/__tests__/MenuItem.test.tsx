import { OrientationMode } from '@enums';
import { MenuItem } from '@models';
import { fireEvent, render, screen } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { describe, expect, it, vi } from 'vitest';

import { MenuListItem } from '../item/MenuListItem';

vi.mock('../sub-menu/SubMenu', () => ({
  SubMenu: ({ children, toggleSubMenu }: PropsWithChildren & { toggleSubMenu: () => void }) => (
    <div data-testid='submenu'>
      <button onClick={toggleSubMenu}>Toggle SubMenu</button>
      {children}
    </div>
  ),
}));

vi.mock('../leaf/MenuLeaf', () => ({
  MenuLeaf: ({ item }: { item: MenuItem }) => <div data-testid='menu-leaf'>{item.label}</div>,
}));

describe('MenuListItem', () => {
  const leafItem = {
    id: '1',
    label: 'Leaf Item',
  };

  const nestedItem = {
    id: '2',
    label: 'Nested Item',
    nestedItems: [
      { id: '3', label: 'Child Item 1' },
      { id: '4', label: 'Child Item 2' },
    ],
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<MenuListItem item={nestedItem} mode={OrientationMode.Horizontal} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders MenuLeaf for leaf nodes', () => {
    render(<MenuListItem item={leafItem} mode={OrientationMode.Horizontal} />);

    const menuLeaf = screen.getByTestId('menu-leaf');
    expect(menuLeaf).toBeInTheDocument();
    expect(menuLeaf).toHaveTextContent('Leaf Item');
  });

  it('renders SubMenu for non-leaf nodes', () => {
    render(<MenuListItem item={nestedItem} mode={OrientationMode.Horizontal} />);

    const subMenu = screen.getByTestId('submenu');
    expect(subMenu).toBeInTheDocument();
  });

  it('toggles SubMenu state when toggleSubMenu is called', () => {
    render(<MenuListItem item={nestedItem} mode={OrientationMode.Horizontal} />);

    const toggleButton = screen.getByText(/toggle submenu/i);
    fireEvent.click(toggleButton);

    // Check if submenu state changes (mocked submenu renders children conditionally)
    expect(screen.getByTestId('submenu')).toBeInTheDocument();
  });

  it('renders nested items recursively', () => {
    render(<MenuListItem item={nestedItem} mode={OrientationMode.Horizontal} />);

    expect(screen.getByText('Child Item 1')).toBeInTheDocument();
    expect(screen.getByText('Child Item 2')).toBeInTheDocument();
  });

  it('passes correct props to SubMenu', () => {
    render(<MenuListItem item={nestedItem} mode={OrientationMode.Horizontal} />);

    const subMenu = screen.getByTestId('submenu');
    expect(subMenu).toBeInTheDocument();
    expect(screen.getByText(nestedItem.nestedItems[0].label)).toBeInTheDocument();
    expect(screen.getByText(nestedItem.nestedItems[1].label)).toBeInTheDocument();
  });

  it('passes correct props to MenuLeaf', () => {
    render(<MenuListItem item={leafItem} mode={OrientationMode.Horizontal} />);

    const menuLeaf = screen.getByTestId('menu-leaf');
    expect(menuLeaf).toHaveTextContent('Leaf Item');
  });
});
