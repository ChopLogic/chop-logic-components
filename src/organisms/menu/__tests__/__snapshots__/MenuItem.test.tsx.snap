// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuListItem > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="submenu"
  >
    <button>
      Toggle SubMenu
    </button>
    <ul
      aria-label="Nested Item"
      class="_subMenuBar_844f10 _subMenuBar__horizontal_844f10"
      role="menu"
    >
      <div
        data-testid="menu-leaf"
      >
        Child Item 1
      </div>
      <div
        data-testid="menu-leaf"
      >
        Child Item 2
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
